### PR TITLE
Add log-softmax and aten::var operations

### PR DESCRIPTION
**Pull request by rdspring1**
Implement the following functions:
1. aten::log_softmax - #1405 
2. aten::_log_softmax_backward_data
3. aten::var
4. aten::std

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp [hunk 1]
```diff
@@ -7,6 +7,64 @@ namespace jit {
...
+    bool unbiased,
+    bool keepdim) {
+  TORCH_INTERNAL_ASSERT(x != nullptr, "Input is invalid.");
+
+  const int kNumberOfDims =
+      TensorDomain::noReductions(x->getMaybeRFactorDomain()).size();
+
+  auto bcast_mean = mean(x, dims, true /* keepdim */);
+  auto x_mean_sub = sub(x, bcast_mean);
+  auto x_mean_sub_sq = mul(x_mean_sub, x_mean_sub);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ops/normalization.cpp:46]: Can you use welford here?

**Review by csarofeen — COMMENTED** [review]
LGTM tagging Jie for parser and type inference changes.

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp (hunk 1 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/ops/normalization.cpp:46]: Yes. I guess when you need both mean and variance, you would be calculating the mean twice.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 2]
```diff
@@ -1830,42 +1833,51 @@ class IrParser {
...
-            MemoryFormat format;
-            std::list<Val*> list_val;
-            std::tie(format, list_val) = getConsistentValues(
-                MemoryFormat::Contiguous(),
-                value_map[node->inputs()[0]->unique()]);
-            auto input_t = list_val.front();
-            list_val.pop_front();
-            auto input = input_t->as<TensorView>();
+      std::array<const char*, kNumSoftmaxFwd> SoftmaxFwd = {
+          "aten::softmax.int(Tensor self, int dim, int? dtype) -> Tensor",
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: nitpick:
"aten::softmax.int(Tensor self, int dim, int? dtype) -> Tensor" could be updated to
"softmax.int(Tensor self, int dim, ScalarType? dtype=None) -> Tensor"

I never really understand their schema matching.... Magically here `int?` and `ScalarType?` resolves to the same thing, but sometimes argument names have to match exactly :shrug:

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 3]
```diff
@@ -1915,35 +1927,94 @@ class IrParser {
...
+              return true;
+            },
+            [](const Node* node) -> OperatorType {
+              return OperatorType::Normalization;
+            });
+      }
+    }
+
+    {
+      std::array<const char*, kNumVarOps> Variance = {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:1970]: var/std performs reduction where output shapes doesn't match input shapes. This means we need to add an entry at shape expression here:https://github.com/csarofeen/pytorch/blob/rds_functorch_ops/torch/csrc/jit/codegen/cuda/graph_fuser.cpp#L941-L965

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:1970]: Note to myself, we should probably have a python tests where we check for existing ops and their shape expression logic.

**Review by jjsjann123 — CHANGES_REQUESTED** [review]
Let's add shape expression logic for added reduction ops.

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp [hunk 4]
```diff
@@ -940,7 +940,11 @@ struct CudaGraphFuser {
       // extended shape expression support to reduction operations
       // TODO: `aten::sum` is too flexible, we should restrict for a better
       // match
-      if (n->kind() == aten::sum) {
+      // TODO: Add python tests where we check for existing ops and their
+      // shape expression logic.
+      static std::unordered_set<Symbol> reduction_ops(
+          {aten::sum, aten::mean, aten::var, aten::std});
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/graph_fuser.cpp:946]: :+1:

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/graph_fuser.cpp:946]: This looks about right. Since we've already have python tests for aten::sum, I don't think it's necessary to duplicate that for the rest of the ops.

**Review by jjsjann123 — APPROVED** [review]
